### PR TITLE
cpuinfo with personality

### PR DESCRIPTION
**Pull request by mihalicyn**
An attempt to address issue https://github.com/lxc/lxcfs/issues/553

### CODE REVIEW

**Comment by mihalicyn** [comment]
Hm, `gettid()` was replaced by `syscall(SYS_gettid)` to make build on Ubuntu 18.04 happy.

File: src/utils.c [hunk 1]
```diff
@@ -642,3 +642,43 @@ DIR *opendir_flags(const char *path, int flags)
...
+	__do_close int fd = -EBADF;
+	int ret = -1;
+	char path[STRLITERALLEN("/proc//personality") + INTTYPE_TO_STRLEN(pid_t) + 1];
+	/* seq_printf(m, "%08x\n", task->personality); */
+	char buf[8 + 1];
+
+	snprintf(path, sizeof(path), "/proc/%d/personality", pid);
+	fd = open(path, O_RDONLY | O_CLOEXEC);
+	if (fd < 0)
+		goto err;
```
Comment (brauner) [inline src/utils.c:657]: I think you can just `return -1` here directly.

File: src/utils.c [hunk 2]
```diff
@@ -642,3 +642,43 @@ DIR *opendir_flags(const char *path, int flags)
...
+
+	ret = read(fd, buf, sizeof(buf) - 1);
```
Comment (brauner) [inline src/utils.c:659]: Please use `read_nointr()` from `utils.{c,h}`.

File: src/utils.c [hunk 3]
```diff
@@ -642,3 +642,43 @@ DIR *opendir_flags(const char *path, int flags)
...
+	if (ret >= 0) {
+		buf[ret] = '\0';
+		*personality = atoi(buf);
```
Comment (brauner) [inline src/utils.c:662]: In LXC we have

```c
int lxc_safe_uint(const char *numstr, unsigned int *converted)
{
        char *err = NULL;
        unsigned long int uli;

        while (isspace(*numstr))
                numstr++;

        if (*numstr == '-')
                return -EINVAL;

        errno = 0;
        uli = strtoul(numstr, &err, 0);
        if (errno == ERANGE && uli == ULONG_MAX)
                return -ERANGE;

        if (err == numstr || *err != '\0')
                return -EINVAL;

        if (uli > UINT_MAX)
                return -ERANGE;

        *converted = (unsigned int)uli;
        return 0;
}
```

apparently I've never ported them to LXCFS.

File: src/proc_cpuview.c [hunk 4]
```diff
@@ -937,6 +966,28 @@ int proc_cpuinfo_read(char *buf, size_t size, off_t offset,
 		return total_len;
 	}
 
+	if (get_host_personality(&host_personality) < 0)
```
Comment (brauner) [inline src/proc_cpuview.c:969]: I don't think we should call this here. Instead we should call this once when the shared module is loaded in the function marked as constructor in `bindings.c`.

So something like the following (UNTESTED):

```diff
diff --git a/src/bindings.c b/src/bindings.c
index 422a622..5ff6a76 100644
--- a/src/bindings.c
+++ b/src/bindings.c
@@ -45,9 +45,15 @@ static bool can_use_swap;
 static bool can_use_sys_cpu;
 static bool has_versioned_opts;
 static bool memory_is_cgroupv2;
+static __u32 host_personality;

 static volatile sig_atomic_t reload_successful;

+__u32 liblxcfs_personality(void)
+{
+       return host_personality;
+}
+
 bool liblxcfs_functional(void)
 {
        return reload_successful != 0;
@@ -919,6 +925,11 @@ static void __attribute__((constructor)) lxcfs_init(void)
                goto broken_upgrade;
        }

+       if (get_host_personality(&host_personality) < 0) {
+               lxcfs_info("Failed to retrieve host personality");
+               goto broken_upgrade;
+       }
+
        reload_successful = 1;
        return;

diff --git a/src/bindings.h b/src/bindings.h
index e35423a..1624784 100644
--- a/src/bindings.h
+++ b/src/bindings.h
@@ -107,6 +107,7 @@ extern bool liblxcfs_can_use_swap(void);
 extern bool liblxcfs_memory_is_cgroupv2(void);
 extern bool liblxcfs_can_use_sys_cpu(void);
 extern bool liblxcfs_has_versioned_opts(void);
+extern __u32 liblxcfs_personality(void);

 static inline bool lxcfs_has_opt(struct lxcfs_opts *opts, lxcfs_opt_t opt)
 {
```

File: src/proc_cpuview.c [hunk 5]
```diff
@@ -937,6 +966,28 @@ int proc_cpuinfo_read(char *buf, size_t size, off_t offset,
...
+		if (ret == -1) {
+			return log_error(0, "Call to personality(%d) failed: %s\n",
+				caller_personality, strerror(errno));
+		}
+
+		lxcfs_debug("task (tid: %d) personality was changed %d -> %d\n",
+				(int)syscall(SYS_gettid), ret, caller_personality);
+
+		personality_changed = true;
+	}
```
Comment (brauner) [inline src/proc_cpuview.c:989]: So I'm not completely sure but I think this doesn't work. We can't change the personality past this function lest we confuse subsequent callers. The thread is a worker thread so doesn't go away. IOW, at the end of this function you need to restore the original personality. One possibility is to wrap `proc_cpuinfo_read()` so am like:

```diff
diff --git a/src/proc_fuse.c b/src/proc_fuse.c
index 5416d43..2e6bc09 100644
--- a/src/proc_fuse.c
+++ b/src/proc_fuse.c
@@ -1513,8 +1513,11 @@ __lxcfs_fuse_ops int proc_read(const char *path, char *buf, size_t size,
                return read_file_fuse_with_offset(LXC_TYPE_PROC_MEMINFO_PATH,
                                                  buf, size, offset, f);
        case LXC_TYPE_PROC_CPUINFO:
-               if (liblxcfs_functional())
+               if (liblxcfs_functional()) {
+                       // assume personality
                        return proc_cpuinfo_read(buf, size, offset, fi);
+                       // restore original personality
+               }

                return read_file_fuse_with_offset(LXC_TYPE_PROC_CPUINFO_PATH,
                                                  buf, size, offset, f);
```

File: src/proc_cpuview.c [hunk 6]
```diff
@@ -920,6 +947,8 @@ int proc_cpuinfo_read(char *buf, size_t size, off_t offset,
 	bool use_view;
 	char *cache = d->buf;
 	size_t cache_size = d->buflen;
+	__u32 host_personality, caller_personality;
+	call_cleaner(restore_personality) bool personality_changed = false;
```
Comment (brauner) [inline src/proc_cpuview.c:951]: Ah, I missed that. I think wrapping `proc_cpuinfo_read()` might be cleaner though.

**Review by brauner — CHANGES_REQUESTED** [review]

File: src/utils.c (hunk 3 above)
Comment (mihalicyn) [reply src/utils.c]: ported, fixed ;-)

File: src/proc_cpuview.c (hunk 4 above)
Comment (mihalicyn) [reply src/proc_cpuview.c]: Thanks! Have done. :-)

File: src/proc_cpuview.c (hunk 6 above)
Comment (mihalicyn) [reply src/proc_cpuview.c]: I've introduced the wrapper function `proc_read_with_personality` for such cases.

File: src/utils.c [hunk 7]
```diff
@@ -642,3 +642,43 @@ DIR *opendir_flags(const char *path, int flags)
...
+
+int get_task_personality(pid_t pid, __u32 *personality)
+{
+	__do_close int fd = -EBADF;
+	int ret = -1;
+	char path[STRLITERALLEN("/proc//personality") + INTTYPE_TO_STRLEN(pid_t) + 1];
+	/* seq_printf(m, "%08x\n", task->personality); */
+	char buf[8 + 1];
+
+	snprintf(path, sizeof(path), "/proc/%d/personality", pid);
```
[the remark this answers is not in the document]
Comment (mihalicyn) [reply src/utils.c]: Agreed totally. I've ported this macro and used it. I'll also walk through other places in lxcfs where we use `snprintf` without checking retvalue and apply this macro in a separate PR.

**Comment by mihalicyn** [comment]
Not related to this PR directly, but thanks to Christian I've noticed that in `lxcfs_init` function we combining `lxc_info` and `lxc_exit` calls on error paths, like this:
```
	init_ns = preserve_ns(pid, "mnt");
	if (init_ns < 0) {
		lxcfs_info("Failed to preserve initial mount namespace");
		goto broken_upgrade;
	}

	/* This function calls unshare(CLONE_NEWNS) our initial mount namespace
	 * to privately mount lxcfs cgroups. */
	if (!cgfs_setup_controllers()) {
		log_exit("Failed to setup private cgroup mounts for lxcfs");
		goto broken_upgrade;
	}
```

It seems strange because `log_exit` implies `exit` call in it. So, we don't reach goto in this case. Looks like a bug and should be addressed in another PR.

File: src/proc_fuse.c [hunk 8]
```diff
@@ -1500,6 +1501,46 @@ static int proc_slabinfo_read(char *buf, size_t size, off_t offset,
...
+				caller_personality, strerror(errno));
+
+		lxcfs_debug("task (tid: %d) personality was changed %d -> %d\n",
+				(int)syscall(SYS_gettid), ret, caller_personality);
+	}
+
+	read_ret = do_proc_read(buf, size, offset, fi);
+
+	if (change_personality) {
+		ret = personality(host_personality);
```
Comment (brauner) [inline src/proc_fuse.c:1532]: We usually just check  `(ret < 0)` instead of equal -1. But it's not that important.

**Review by brauner — APPROVED** [review]

**Pull request merged**